src/py/flwr/server/superlink/state/in_memory_state.py

```diff
@@ -117,6 +117,35 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
             log(ERROR, errors)
             return None
 
+        with self.lock:
+            # Check if the TaskIns it is replying to exists and is valid
+            task_ins_id = task_res.task.ancestry[0]
+            task_ins = self.task_ins_store.get(UUID(task_ins_id))
+
+            if task_ins is None:
+                log(ERROR, "TaskIns with task_id %s does not exist.", task_ins_id)
+                return None
+
+            if task_ins.task.created_at + task_ins.task.ttl <= time.time():
+                log(ERROR, "TaskIns with task_id %s is expired.", task_ins_id)
+                return None
+
+            # Limit the TaskRes TTL to not exceed the
+            # expiration time of the TaskIns it replies to.
+            # Condition: TaskIns.created_at + TaskIns.ttl ≥
+            #            TaskRes.created_at + TaskRes.ttl
+            if (
+                task_res.task.ttl
+                > task_ins.task.created_at
+                + task_ins.task.ttl
+                - task_res.task.created_at
+            ):
+                task_res.task.ttl = (
+                    task_ins.task.created_at
+                    + task_ins.task.ttl
+                    - task_res.task.created_at
+                )
```

Comment:
I think we should log a warning/info if the `SuperLink` receives a `TaskRes` with larger than expected TTL and updates it.